### ISSUE TITLE
offsetof issue ?

### ISSUE THREAD

**Issue by oandrieu** [AbsInt/CompCert#47]
I cannot get the `offsetof` macro to work …

```c
#include <stddef.h>

struct toto {
  int x;
  int y;
};

int main()
{
  return offsetof(struct toto, y);
}
```
results in:
```
$ ccomp -c off2.c 
off2.c:10: syntax error: expecting ')'.
off2.c:10: this is the location of the unclosed '('.
Fatal error; compilation aborted.
1 error detected.
```

but the parentheses seem balanced in the preprocessor output:
```
$ ccomp -E off2.c | grep return
  return ((size_t) &(((struct toto)*) 0)->y);
```

**Comment by oandrieu** [comment]
Right, so yes, I think this is an issue in the definition of `offsetof` in `stddef.h`.
The expansion contains a cast expression: `((ty)*) NULL` but `(ty)*` is not a valid `type-name` fragment, the inner parentheses shouldn't be there.
The error message is a bit confusing though.

```diff
diff --git a/runtime/include/stddef.h b/runtime/include/stddef.h
index 290434f..575b89b 100644
--- a/runtime/include/stddef.h
+++ b/runtime/include/stddef.h
@@ -98,7 +98,7 @@ typedef signed int wchar_t;
 #endif
 
 #if defined(_STDDEF_H) && !defined(offsetof)
-#define offsetof(ty,member) ((size_t) &(((ty)*) NULL)->member)
+#define offsetof(ty,member) ((size_t) &((ty *) NULL)->member)
 #endif
 
 #endif
```

**Comment by bschommer** [comment]
Thanks for the hint. Should be fixed by commit ef92c55cf1f023ec43058eaaf2b176e16fe41fcf
